design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,284 @@
+- providing a SQL-driven testing tool will mesh better with our product offering, allowing users to
+    write tests without every leaving the "SQL mindset"
+    
+## Scope
+
+This KLIP covers:
+
+- design of the test file format
+- design of the testing-specific directives
+- outline features that need to be implemented in order to replace (R)QTT/testing-tool
+
+This KLIP does not cover:
+
+- implementation of a migration path from (R)QTT/testing-tool, but we plan to deprecate and remove
+    the ksql-test-runner over the next few releases. At the time when we remove it, we will build
+    some convenience scripts to help generate YATT sql files.
+- generating historical plans from YATT sql tests
+
+The intention of this KLIP is to provide a solid foundation for YATT to eventually replace the other
+testing tools and to make sure we are working towards a cleaner code base instead of littering
+it with testing code that needs to be maintained and will eventually slow down our development
+velocity. That being said, it does not cover all the details required, or provide a timeline, to
+migrate the existing tools.
+
+## Design
+
+The best way to describe the testing tool is by a motivating example, such as the test file below:
+
+```sql
+---------------------------------------------------------------------------------------------------
+--@test: dml - stream - add filter
+---------------------------------------------------------------------------------------------------
+
+CREATE STREAM foo (id VARCHAR KEY, col1 INT) WITH (kafka_topic='foo', value_format='JSON');
+CREATE STREAM bar AS SELECT * FROM foo;
+
+INSERT INTO foo (rowtime, id, col1) VALUES (1, '1', 1);
+
+ASSERT DATA bar OFFSET 1 (rowtime, id, col1) VALUES (1, '1', 1);
+
+CREATE OR REPLACE STREAM bar AS SELECT * FROM foo WHERE col1 = 123;
+
+INSERT INTO foo (rowtime, id, col1) VALUES (2, '2', 2);
+INSERT INTO foo (rowtime, id, col1) VALUES (3, '3', 123);
+
+ASSERT DATA bar OFFSET 2 (rowtime, id, col1) VALUES (3, '3', 123);
+ASSERT STREAM bar (id VARCHAR KEY, col1 INT) WITH (kafka_topic='BAR', value_format='JSON');
+
+---------------------------------------------------------------------------------------------------
+--@test: dml - stream - change column
+--@expected.error: io.confluent.ksql.util.KsqlException
+--@expected.message: "Cannot REPLACE data source: DataSource '`BAR`' has schema ..."
+---------------------------------------------------------------------------------------------------
+
+CREATE STREAM foo (id VARCHAR KEY, col1 VARCHAR, col2 VARCHAR) 
+    WITH (kafka_topic='foo', value_format='JSON');
+
+-- the below operation is invalid because "col2" will be missing from the
+-- schema after renaming it to col3
+CREATE STREAM bar AS SELECT id, col1, col2 FROM foo;
+CREATE OR REPLACE STREAM bar AS SELECT id, col1, col2 AS col3 FROM foo;
+```
+
+### Test Structure
+
+YATT will accept as a parameter a directory containing testing files, and will run all the
```

Comment:
Should we also support passing in a single test file directly, so users can debug individual tests/test files without having to run an entire directory of tests?

Reply:
yup! I will add this into the KLIP